src/js/actions/export.js

```diff
@@ -0,0 +1,222 @@
+        documentLib = require("adapter/lib/document"),
+        layerLib = require("adapter/lib/layer");
+
+    var EXTENSION_DATA_NAMESPACE = "designSpace";
+
+    /**
+     * Event handlers initialized in beforeStartup.
+     *
+     * @private
+     * @type {function()}
+     */
+    var _exportService;
+
+    /**
+     * before start up
+     * @return {Promise}
+     */
+    var beforeStartup = function () {
+        _exportService = new ExportService();
+        return _exportService.init();
+    };
+    beforeStartup.reads = [];
+    beforeStartup.writes = [];
+
+    var openExportPanel = function () {
+        return this.transfer(dialog.openDialog, "exports-panel-dialog");
+    };
+    openExportPanel.reads = [];
+    openExportPanel.writes = [locks.JS_DIALOG];
+
+
+    var _exportAsset = function (documentID, layer, asset) {
+        return _exportService.sendMessage("exportLayer", { layer: layer, scale: asset.scale })
+            .bind(this)
+            .then(function (exportResponse) {
+                // TODO we should inspect the response a little more closely, for export errors
+                if (exportResponse && Array.isArray(exportResponse.result)) {
+                    var assetProps = asset.toJS(),
+                        layerExportsMap = new Map(),
+                        payload;
+
+                    assetProps.filePath = exportResponse.result[0];
+                    assetProps.status = "stable"; // FIXME need a better way to handle these constants
+                    layerExportsMap.set(layer.id, assetProps);
+
+                    payload = {
+                        documentID: documentID,
+                        documentExports: {
+                            layerExportsMap: layerExportsMap
+                        }
+                    };
+                    this.dispatch(events.export.ASSET_CHANGED, payload);
+
+                    var documentExports = this.flux.stores.export.getDocumentExports(documentID),
+                        layerExportsArray = documentExports && documentExports.layerExportsArray(layer.id);
+
+                    return _updateLayerExportMetadata(documentID, layer.id, layerExportsArray);
+                } else {
+                    return Promise.resolve();
+                }
+            });
+    };
+
+    var exportAllAssets = function () {
+        var document = this.flux.stores.application.getCurrentDocument();
+
+        if (!document) {
+            Promise.resolve("No Document");
+        }
+
+        var documentID = document.id,
+            documentExports = this.flux.stores.export.getDocumentExports(documentID),
+            layerExportsMap = documentExports && documentExports.layerExportsMap,
+            exportArray = [];
+
+        if (!layerExportsMap || layerExportsMap.size < 1) {
+            return Promise.resolve("no assets to export");
+        }
+
+        layerExportsMap.forEach(function (layerExportAssets, layerID) {
```

Comment:
Aside: the worst part of `window.Map` is the dearth of built-in iterators compared to those available in Immutable, like `.reduce`, `.map`, `.flatMap`, etc. (I think this would be nicer with the latter two.) We could always just write those helper functions ourselves though and stick them on `window.Map.prototype`. Food for thought :poultry_leg: